toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -110,13 +119,85 @@ func fetchPackages() (err error) {
+		err = downloadDeltaNodes(dependencyGraph, cloner)
+		if err != nil {
+			err = fmt.Errorf("failed to download delta RPMs: %w", err)
+			return
+		}
+		// Update the package graph with the paths to the delta RPMs we downloaded
+		err = pkggraph.WriteDOTGraphFile(dependencyGraph, *outputGraph)
+		if err != nil {
+			err = fmt.Errorf("failed to write cache graph to file: %w", err)
+			return
+		}
+	}
+
 	// If we grabbed any RPMs, we need to convert them into a local repo
-	if hasUnresolvedNodes {
+	if *tryDownloadDeltaRPMs || hasUnresolvedNodes {
+		logger.Log.Info("Configuring downloaded RPMs as a local repository")
 		err = cloner.ConvertDownloadedPackagesIntoRepo()
 		if err != nil {
 			err = fmt.Errorf("failed to convert downloaded RPMs into a repo:\n%w", err)
 			return err
 		}
+
+		if strings.TrimSpace(*outputSummaryFile) != "" {
+			err = repoutils.SaveClonedRepoContents(cloner, *outputSummaryFile)
+			if err != nil {
+				logger.Log.Errorf("Failed to save cloned repo contents.")
+				return
+			}
+		}
+	}
+	return
+}
+
+// downloadDeltaNodes will look at the final cached graph we saved and see if any RPMS can be download instead of built.
+// If the previous part of the fetcher worked well we should be able to download only the delta RPMs we need
+// to build our packages or image (i.e. we should be able to create a subgraph just like we would for the build step)
+//   - dependencyGraph: The graph to use to find the packages we need to build. Should have any caching operations already
+//     performed on it. Will be updated with the paths to the delta RPMs we download.
+//   - cloner: The cloner to use to download the RPMs
+func downloadDeltaNodes(dependencyGraph *pkggraph.PkgGraph, cloner *rpmrepocloner.RpmRepoCloner) (err error) {
+	const (
+		deltaBuildGraph = true
+	)
+
+	timestamp.StartEvent("delta package download", nil)
+	defer timestamp.StopEvent(nil)
+
+	// Generate the list of packages that need to be built. If none are requested then all packages will be built. We
+	// don't care about explicit rebuilds here since we are going to rebuild them anyway.
+	packageVersToBuild, err := parseAndGeneratePackageList(dependencyGraph, *pkgsToBuild, *pkgsToRebuild, *pkgsToIgnore, *imageConfig, *baseDirPath)
+	if err != nil {
+		err = fmt.Errorf("unable to generate package build list to calculate delta downloads: %w", err)
+		return
+	}
+
+	timestamp.StartEvent("graph initialization", nil)
+	// The scheduler utils expect to pick a graph up from a file, so we will write the graph we wrote it to a file and
+	// now we read it back in and optimize it. We will heavily modify this graph so it should not be used for anything
+	// else.
+	isGraphOptimized, deltaPkgGraphCopy, _, err := schedulerutils.InitializeGraph(*outputGraph, packageVersToBuild, deltaBuildGraph)
+	if err != nil {
+		err = fmt.Errorf("failed to initialize graph for delta package downloading: %w", err)
+		return
+	}
+	timestamp.StopEvent(nil) // graph initialization
```

Comment:
>timestamp.StopEvent(nil) // graph initialization [](http://example.com/codeflow?start=1&length=48)

This is problematic, since we can now return before we stop the event. I'd:
- put each start-stop scope into a separate function, or
- put all stops into one `defer` call, or
- check how the parenting works for these events (the `nil` argument we pass everywhere) - I'm not clear on that one, but maybe there's a way to hook them to each other, so we only need to remember to stop the top parent.

Also, I'm not sure, if we wouldn't need a re-design for the timestamps, but that's out of scope of this PR.

Reply:
Another thought: we could just move the time monitoring inside `InitializeGraph`. This should solve the issue of the `defer` call.

Reply:
The timestamps work best at the scope of a function. The `nil` is normal, we only need to pass an explicit value around when dealing with multithreaded workers.

Also, in general we make no guarantees about the quality of the timestamp data after an error (the timestamps will automatically close out any lingering measurements when the stop).

Reply:
Just moved it into the initialize function though, since that makes sense.